src/axom/core/tests/core_benchmark_array.cpp

```diff
@@ -0,0 +1,357 @@
+// Copyright (c) 2017-2024, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include "benchmark/benchmark.h"
+#include "axom/config.hpp"
+#include "axom/core.hpp"
+#include "axom/slic.hpp"
+#include "axom/fmt.hpp"
+
+#include <tuple>
+#include <vector>
+#include <type_traits>
+#include <cstdint>
+
+//-----------------------------------------------------------------------------
+// Arguments and types to test
+//-----------------------------------------------------------------------------
+
+void CustomArgs(benchmark::internal::Benchmark* b)
+{
+  // b->Arg(1 << 3);   // small
+  b->Arg(1 << 16);  // larger than  32K L1 cache
+  // b->Arg(1 << 19);  // larger than 256K L2 cache
+  // b->Arg(1 << 25);  // larger than  25M L3 cache
+}
+
+template <typename T>
+struct Wrapper
+{
+  T j;
+  bool operator==(const Wrapper& other) const { return j == other.j; }
+};
+
+using Types = std::tuple<int, std::pair<int, int>, Wrapper<int>, std::string>;
+
+//-----------------------------------------------------------------------------
+// Helper functions for generating type names and values
+//-----------------------------------------------------------------------------
+
+template <typename T>
+struct get_value_impl
+{
+  static T get(int i) { return static_cast<T>(i); }
+};
+
+template <typename T>
+struct get_value_impl<std::pair<T, T>>
+{
+  static std::pair<T, T> get(int i)
+  {
+    return std::make_pair(static_cast<T>(i), static_cast<T>(i + 1));
+  }
+};
+
+template <typename T>
+struct get_value_impl<Wrapper<T>>
+{
+  static Wrapper<T> get(int i) { return Wrapper<T> {static_cast<T>(i)}; }
+};
+
+template <>
+struct get_value_impl<std::string>
+{
+  static std::string get(int i)
+  {
+    return axom::fmt::format("This is a somewhat long string -- {}", i);
+  }
+};
+
+template <typename T>
+T get_value(int i)
+{
+  return get_value_impl<T>::get(i);
+}
+
+template <typename T>
+std::string get_type_name();
+
+template <>
+std::string get_type_name<int>()
+{
+  return "int";
+}
+
+template <>
+std::string get_type_name<std::pair<int, int>>()
+{
+  return "std::pair<int, int>";
+}
+
+template <>
+std::string get_type_name<Wrapper<int>>()
+{
+  return "Wrapper<int>";
+}
+
+template <>
+std::string get_type_name<std::string>()
+{
+  return "std::string";
+}
+
+// Helper function to check that an array's size and capacity match expectation
+// Since we're benchmarking, let's do our best to ensure it's a no-op in release configs
+template <typename Arr>
+void assert_size_and_capacity(const Arr& arr, int exp_size, int exp_capacity)
+{
+#ifdef NDEBUG
```

Comment:
Do we have a coding standard to prefer '#if defined(...)' instead?

Reply:
Not that I'm aware of, and a quick check of the coding guidelines didn't turn up anything. Should we prefer `#if defined()`?

Reply:
I prefer it since it's more flexible. However, it my not matter to most folks.